Security v2

val-metal opened the pull request.

val-metal committed:
security

kozusznik commented on dotaz.py:
Do not use Czech language for naming files or identifier.

kozusznik commented on doc/SECURITY.md:
```diff
@@ -0,0 +1,19 @@
+Working bunch of parameters for running from IDE:
```
Improve documentation, please

kozusznik commented on dotaz.py:
```diff
@@ -0,0 +1,20 @@
+import requests
```
Consider moving of this tests into src/test directory from project root

kozusznik commented on dotaz.py:
```diff
@@ -0,0 +1,20 @@
+import requests
+import json
```
consider transform this into JUnit test using 
Apache HttpClient library:
https://mkyong.com/java/apache-httpclient-examples/

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/Constants.java:
```diff
@@ -17,7 +17,7 @@ final class Constants {
-	static final String SECURITY_SERVERS =
-		"cz.it4i.fiji.datastore.security.servers";
+	static final String SECURITY_SERVERS = "cz.it4i.fiji.datastore.security.servers";
+	static final String SECURITY_ABLED ="cz.it4i.fiji.datastore.security.able";
```
add empty line.
Consider using name SECURITY_ON

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerEndpoint.java:
```diff
@@ -0,0 +1,89 @@
+import com.google.gson.Gson;
+
+import javax.inject.Inject;
+        import javax.ws.rs.*;
```
use formatter with style for scijava

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerEndpoint.java:
```diff
@@ -0,0 +1,89 @@
+
+    //@Produces(MediaType.APPLICATION_JSON)
+    public Response getAllOAuthServers() {
+        String vysledek="";
```
do not use Czech language for identifier names.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerNew.java:
```diff
@@ -0,0 +1,49 @@
+@Table(name = "oauth_server")
+@Getter
+@Setter
+public class OAuthServerNew extends PanacheEntityBase {
```
what does mean a suffix "new"

kozusznik replied:
Consider using existing class User.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerNew.java:
```diff
@@ -0,0 +1,49 @@
+        this.link_token = link_token;
+    }
+
+    public String BCSerlialize(){
```
what does mean name BCSerlialize
consider using java code conventions:
https://www.oracle.com/technetwork/java/codeconventions-150003.pdf

method identifiers starts with small letter.

consider override to toString method.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerNew.java:
```diff
@@ -0,0 +1,49 @@
+
+    private String link_token;
+
+    public OAuthServerNew()
```
consider using lombok annotation AllArgsConstructor, NoArgCOnstructor.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerService.java:
```diff
@@ -0,0 +1,55 @@
+    public List<OAuthServerNew> getAllOAuthServers() {
+        if(entityManager==null)
+        {
+            return new ArrayList<OAuthServerNew>();
```
consider using Collections.emptyList

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthServerService.java:
```diff
@@ -0,0 +1,55 @@
+        {
+            return new ArrayList<OAuthServerNew>();
+        }
+        return entityManager.createQuery("SELECT os FROM OAuthServerNew os", OAuthServerNew.class)
```
consider using OAuthServerNew.findAll

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserEndpoint.java:
```diff
@@ -0,0 +1,75 @@
+        List<OAuthUserNew> list=oauthUserService.getAllOAuthUsers();
+        for(OAuthUserNew oa :list)
+        {
+            //TODO Anonimize tokens
```
consider setting a null value.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserNew.java:
```diff
@@ -0,0 +1,41 @@
+package cz.it4i.fiji.datastore.security;
```
same comments as for OAuthServer

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserService.java:
```diff
@@ -0,0 +1,48 @@
+
+    public List<OAuthUserNew> getAllOAuthUsers() {
+
+         return entityManager.createQuery("SELECT os FROM OAuthUserNew os", OAuthUserNew.class)
```
consider using OAuthUser.findAll()

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityEndpoint.java:
```diff
@@ -51,9 +52,20 @@ public Response get(@Context UriInfo requestURI, @Context HttpHeaders headers,
 
 	}
 
+	@GET
```
it should be done automatically after changes.

val-metal committed:
add OAuthGroups

Code cleaning still needs to be done, it will probably happen tomorrow

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthGroup.java:
```diff
@@ -0,0 +1,86 @@
+    private int id;
+    private String name;
+    private int ownerID;
+    @ElementCollection
```
There is User and Dataset so use association instead of ElementCollection.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthGroupEndpoint.java:
```diff
@@ -0,0 +1,108 @@
+    OAuthGroupService oAuthGroupService;
+
+    @POST
+    public Response createOAuthGroup(OAuthGroup group) {
```
use DTO object in REST controller

kozusznik replied:
Consider using http://modelmapper.org for mapper creation.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserEndpoint.java:
```diff
@@ -0,0 +1,75 @@
+            oa.setClientID("censored");
+            oa.setClientSecret("censored");
+        }
+        Gson gson = new Gson();
```
Consider using already used mechanism when DTO objrecy is returned and automacally converted to JSON

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserEndpoint.java:
```diff
@@ -0,0 +1,75 @@
+        return oauthUserService.getAllOAuthUsers();
+    }
+    @GET
+    @Path("/json")
```
Consider using standard convention for forming REST API - requested type of data probably should not be part of a path.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserEndpoint.java:
```diff
@@ -0,0 +1,75 @@
+
+
+    @GET
+    @Path("/{client_id}")
```
Consider using constants for naming parameters in HTTP path

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/PermissionType.java:
```diff
@@ -0,0 +1,23 @@
+    R("r"), // Read
+    W("w"), // Write
+    C("c"), // Create
+    RW("rw"); // Read-Write
```
should be used with EnumSet when we use Read in conjuction with write.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityEndpoint.java:
```diff
@@ -38,11 +36,17 @@ public class SecurityEndpoint {
+
 	public Response get(@Context UriInfo requestURI, @Context HttpHeaders headers,
-		@PathParam(OAUTH_SERVER) String oauthServer, @QueryParam(CODE) String code)
+		@PathParam(OAUTH_SERVER) String oauthServer, @QueryParam(CODE) String code,String device_id,String device_name)
```
how is binded device_id and device_name and what does it mean?

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityRegistry.java:
```diff
@@ -34,7 +32,66 @@ class SecurityRegistry {
+	private Boolean staticSecurity;
+
+
+	public List<String> getServerAliases()
```
servers should be also persisted in DB.

kozusznik commented on start-server-https.cmd:
```diff
@@ -0,0 +1,17 @@
+set myVar="C:\Users\ValMe\Documents\GitHub\hpc-datastore"
```
enviromental dependent information should not be in git.
Similar thinks is solved that you create template and commit it. Whilst file that should is derived from a template is mentioned in .gitignore.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityRegistry.java:
```diff
@@ -34,7 +32,66 @@ class SecurityRegistry {
+		LinkedList<OAuthServer> temp = new LinkedList<>();
+
+		if(isStaticSercuritySet()) {
+			String[] serversTokens = getProperty(SECURITY_SERVERS).split(";");
```
Use security properties for an initialisation of an empty database. After that ignore it. For reading data use only db.

kozusznik commented on myDb.mv.db:
Consider remove data from git and add its file name to gitignore.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/GroupsAuthentication.java:
How it works? Something missing for me.

val-metal committed:
Updates and improvements

Fixed most of the problem from the previous PR

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityRegistry.java:
```diff
@@ -99,54 +173,150 @@ private void initServers() {
+			{
+				continue;
+			}
+			List<Integer> userids=groups.get(i).getUsersIDs();
```
Compilation error

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityRegistry.java:
```diff
@@ -99,54 +173,150 @@ private void initServers() {
+			List<Integer> userids=groups.get(i).getUsersIDs();
+			for(int j=0;j<userids.size();j++)
+			{
+				String serialized_user=user.get(userids.get(j)).BCSerlialize_user();
```
Compilation error.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/SecurityRegistry.java:
```diff
@@ -99,54 +173,150 @@ private void initServers() {
+		list_all_users = OAuthUserNew.listAll();
+
+		for (int i = 0; i < list_all_users.size(); i++) {
+			list_user.add(list_all_users.get(i).BCSerlialize_user().split(";"));
```
compilation error

val-metal commented:
To je zvláštní. Zkusím se na to co nejdříve podívat.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserNew.java:
```diff
@@ -0,0 +1,37 @@
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class OAuthUserNew extends PanacheEntityBase {
```
Is PanacheEntityRepository necessary? I think, there is used only EntityManager.

kozusznik commented on src/main/java/cz/it4i/fiji/datastore/security/OAuthUserNew.java:
```diff
@@ -0,0 +1,37 @@
+    private String OAuthAlias;
+    private String ClientToken;
+    private String ClientID;
+    private String ClientSecret;
```
Respect Code Convention according to names of Java identifiers 
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

val-metal committed:
working updated

working ,tests and rest of the updates will come tomorow

val-metal committed:
Almost working added tests+UI

Project is almost functional only groups need some touches +added tests and simple UI for CRUD on OAuth entities trought console.

val-metal committed:
repairs
